rfcs/syntax-type-alias-type-packs.md

```diff
@@ -0,0 +1,178 @@
+Type pack parameters after the first one have to be type packs:
+```lua
+type Y<T..., U...> = --
+
+type A<S...> = Y<S..., S...>              -- T... = (S...), U... = (S...)
+type B<S...> = Y<number, ...string, S...> -- T... = (number, ...string), U... = S...
+type C<S...> = Y<number, string, S...>    -- error, T... = (number, string, S...), but U... = undefined
+type D = Y<...number>                     -- error, T = (...number), but U... = undefined, not (...number) even though one infinite set is enough to fill two, we may have '...number' inside a type pack argument and we'll be unable to see its content
+
+type Z<T, U...> = --
+
+type E<S...> = Z<number, S...>         -- T = number, U... = (S...)
+type F<S...> = Z<number, string, S...> -- T = number, U... = (string, S...)
+type G<S...> = Z<S...>                 -- error, not enough regular type arguments, can't split the front of S... into T
+
+type W<T, U..., V...> = --
+
+type H<S..., R...> = W<number, S..., R...>         -- U... = S..., V... = R...
+type I<S..., R...> = W<number, string, S..., R...> -- U... = (string, S...), V... = R...
+```
+
+### Explicit type pack syntax
+
+To enable additional control for the content of a type pack, especially in cases where multiple type pack parameters are expected, we introduce an explicit type pack syntax for use in type alias instantiation.
+
+Similar to variadic types `...a` and generic type packs `T...`, explicit type packs can only be used at type pack positions:
+```lua
+type Y<T..., U...> = (T...) -> (U...)
+
+type F1 = Y<(number, string), (boolean)>        -- T... = (number, string), U... = (boolean)
+type F2 = Y<(), ()>                             -- T... = (), U... = ()
+type F3<S...> = Y<string, S..., (number, S...)> -- T... = (string, S...), U... = (number, S...)
+```
+
+In type parameter list, types inside the parentheses always produce a type pack.
+This is in contrast to function return type pack annotation, where `() -> number` is the same as `() -> (number)`.
+
+This is a breaking change.
+
+Users can already have type alias instantiations like these:
+```lua
+type X<T> = T?
+type A = X<(number)> -- valid right now, typechecking error after this RFC
+```
+
+Explicit type pack syntax is not available in other type pack annotation contexts.
+
+## Drawbacks
+
+### Type pack element extraction
+
+Because our type alias instantiations are not lazy, it's impossible to split of a single type from a type pack:
+```lua
+type Car<T, U...> = T
+
+type X = Car<number, string, boolean> -- number
+type Y<S...> = Car<S...>              -- error, not enough regular type arguments
+type Z = Y<number, string, boolean>   -- error, Y doesn't have a valid definition
+```
+
+Splitting off a single type is is a common pattern with variadic templates in C++, but we don't allow type alias overloads, so use cases are more limited.
```

Comment:
I find explicitly representing packs this way to be a surprising direction, but that might just be my familiarity with C++ templates talking. It'd be nice if the rationale for doing it this way was explained a bit more, instead of only how it works mechanically.

It'd still be possible to implement splitting in the future through a magic utility type, right? Something like `Select<T..., number>`.